### PR TITLE
ghc: update bootstrap 8.6.3 -> 8.6.5, add aarch64 support

### PR DESCRIPTION
**Pull request by thefloweringash**
###### Motivation for this change

GHC on Aarch64.

I'm not in the loop with how the haskell infrastructure works in nixpkgs, I just require ghc to build my system for things like pandoc and nix-diff. I assume aarch64 was omitted from ghc863Binary due to the [upstream 8.6.3 release][] not to having an appropriate tarball. The [upstream 8.6.5 release][] includes `ghc-8.6.5-aarch64-ubuntu18.04-linux.tar.xz`.

[upstream 8.6.3 release]: https://downloads.haskell.org/~ghc/8.6.3 
[upstream 8.6.5 release]: https://downloads.haskell.org/~ghc/8.6.5 

With these changes, I can build ghc 8.8.3 on aarch64. I split out the commit for the version change and the aarch64 support to make it easier to read.

Related: #80355 -- contains the same llvm logic as here for `propagatedBuildInputs`
Related: #80176 -- issue for this and #80355

cc @marcweber @kosmikus @peti (maintainers of ghc)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thefloweringash** [comment]
I meant to open this as a draft.

**Comment by cdepillabout** [comment]
@thefloweringash This seems reasonable.  Is there any reason you've marked this as a WIP?  Is there anything that is not currently working?

@GrahamcOfBorg build ghc

**Comment by thefloweringash** [comment]
I'm not aware of anything not working. I built ghc, I'll see if I can build pandoc. I'll remove the WIP and DRAFT markers.

There was some contention about the llvm input, see https://github.com/NixOS/nixpkgs/pull/80355#pullrequestreview-364288760 and further comments.

Now that I've seen the build count, this probably shouldn't go to master. Should this go to `staging`, or `haskell-updates`?

**Comment by cdepillabout** [comment]
It looks like `ghc` was able to be built with the new bootstrap GHC, so I'd say this looks good.

Let's see if lens can be built as well:

@GrahamcOfBorg build haskellPackages.lens

---------------------------------------

@thefloweringash Okay, that sounds good.

I think you're right, this should probably go to the `haskell-updates` branch?  Can you do that rebase?

After that, how about giving this 1 week for other people to review it, and if nobody has anything bad to say about it, I'll merge it in.  Please ping me in 1 week if there has been no activity.

Also, cc @matthewbauer and @Ericson2314, who occasionally contribute to the GHC derivation.

File: pkgs/development/compilers/ghc/8.6.5-binary.nix [hunk 1]
```diff
@@ -24,27 +25,33 @@ let
...
     x86_64-darwin = {
       url = "http://haskell.org/ghc/dist/${version}/ghc-${version}-x86_64-apple-darwin.tar.xz";
-      sha256 = "1hbzk57v45176kxcx848p5jn5p1xbp2129ramkbzsk6plyhnkl3r";
+      sha256 = "0s9188vhhgf23q3rjarwhbr524z6h2qga5xaaa2pma03sfqvvhfz";
     };
   }.${stdenv.hostPlatform.system}
     or (throw "cannot bootstrap GHC on this platform"));
 
   nativeBuildInputs = [ perl ];
+  propagatedBuildInputs = lib.optional stdenv.isAarch64 [ llvm ];
```
Comment (cdepillabout) [inline pkgs/development/compilers/ghc/8.6.5-binary.nix]: I think the big question from https://github.com/NixOS/nixpkgs/pull/80355#pullrequestreview-364288760 is, "Why is this needed?"

It seems strange that aarch64 would need this, but nothing else would.  Additionally, why is this not needed for the non-bootstrap GHCs?

I don't really have any knowledge of how GHC is built, or how aarch64 works, but this just _seems_ like the wrong solution.

Comment (thefloweringash) [reply pkgs/development/compilers/ghc/8.6.5-binary.nix]: I believe I have found the answers.

GHC on non-x86 uses the LLVM codegen by default [citation needed]. When using the LLVM codegen, ghc needs llvm binaries on the path.

> You must install and have LLVM available on your `PATH` for the LLVM code generator to work. Specifically GHC needs to be able to call the `opt` and `llc` tools. 

-- (from [10.10.2. LLVM Code Generator (-fllvm)][ghc -fllvm])

[ghc -fllvm]: https://downloads.haskell.org/~ghc/8.6.5/docs/html/users_guide/codegens.html#llvm-code-generator-fllvm

> It seems strange that aarch64 would need this, but nothing else would. Additionally, why is this not needed for the non-bootstrap GHCs?

It turns out this is required for the non-bootstrap GHCs, and is already present in `propagatedBuildInputs` of, for example, ghc 8.8.3:
https://github.com/NixOS/nixpkgs/blob/bfb747aacf5bccf1609170839d63e519cf63ca0c/pkgs/development/compilers/ghc/8.8.3.nix#L193-L194

I've updated the branch to more closely resemble the non bootstrap GHC structure.

**Review by Ericson2314 — APPROVED** [review]

**Comment by domenkozar** [comment]
@GrahamcOfBorg build ghc

**Review by cdepillabout — APPROVED** [review]

**Comment by thefloweringash** [comment]
The effects of this change can be tracked in [hydra evaluation 1576936][].

[hydra evaluation 1576936]: https://hydra.nixos.org/eval/1576936

**Comment by vcunat** [comment]
Well, down-side is that this jobset only uses `x86_64-linux` :-)
